cpmpy/transformations/linearize.py

```diff
@@ -247,185 +219,52 @@ def only_positive_bv(cpm_expr):
-            # BV1 >=< ~BV2 => BV1 + BV2 >=< 1
-            if isinstance(rhs, NegBoolView):
-                return [Comparison(cpm_expr.name, lhs + rhs._bv, 1)]
-            return [cpm_expr]
+        new_cons = []
 
         if isinstance(lhs, _NumVarImpl):
-            return [cpm_expr]
-
-        if isinstance(lhs, Operator) and lhs.name == "sum" or lhs.name == "wsum":
-            # sum, wsum
-            if lhs.name == "sum":
-                if not any(isinstance(arg, NegBoolView) for arg in lhs.args):
-                    # No NegBoolViews in sum, return
-                    return [cpm_expr]
-                # Convert to wsum
-                lhs = Operator("wsum", [[1] * len(lhs.args), lhs.args])
-
-            if lhs.name == "wsum":
-                # TODO: can be optimized with sum/wsum improvements, see: https://github.com/CPMpy/cpmpy/issues/97
-                weights, vars = [],[]
-                for w, arg in zip(*lhs.args):
-                    if isinstance(arg, NegBoolView):
-                        weights += [w,-w]
-                        vars += [1, arg._bv]
-                    else:
-                        weights += [w]
-                        vars += [arg]
-                new_lhs = Operator("wsum", [weights, vars])
-
-            if isinstance(rhs, NegBoolView):
-                new_lhs += 1 * rhs._bv
-                new_rhs = 1
-            else:
-                new_rhs = rhs
-
-            return [Comparison(cpm_expr.name, new_lhs, new_rhs)]
-
-        if isinstance(lhs, Operator):
-            # min, max
-            simplied_args = [simplify(arg) for arg in lhs.args]
-            nn_args, cons_l = zip(*[get_or_make_var(arg) for arg in simplied_args])
-
-            nn_lhs = Operator(lhs.name, nn_args)
-            nn_rhs, cons_r = get_or_make_var(simplify(rhs))
-
-            var_cons = [c for con in cons_l for c in con] + [c for con in cons_r for c in con]
-            return var_cons + [Comparison(cpm_expr.name, nn_lhs, nn_rhs)]
-
-        if isinstance(lhs, GlobalConstraint):
-            # min, max
-            simplied_args = [simplify(arg) for arg in lhs.args]
-            nn_args, cons_l = zip(*[get_or_make_var(arg) for arg in simplied_args])
-
-            nn_lhs = type(lhs)(nn_args)
-            nn_rhs, cons_r = get_or_make_var(simplify(rhs))
-
-            var_cons = [c for con in cons_l for c in con] + [c for con in cons_r for c in con]
-            return var_cons + [Comparison(cpm_expr.name, nn_lhs, nn_rhs)]
-
-
-    if isinstance(cpm_expr, Operator) and cpm_expr.name == "->":
+            if isinstance(lhs,NegBoolView):
+                lhs, rhs = -lhs._bv, rhs - 1
+
+        elif lhs.name == "sum":
+            weights = [-1 if isinstance(a, NegBoolView) else 1 for a in lhs.args]
+            lhs = sum(w * a._bv if isinstance(a, NegBoolView) else w * a for w,a in zip(weights,lhs.args))
+            rhs -= sum(w < 0 for w in weights)
```

Comment:
a bit obscure... isn't a generic function possible that gets w_lhs, w_exp, rhs as input and outputs the same but without negboolviews?